### DISCUSSION TITLE
How to iterate over the items in a sorted DataGrid?

### DISCUSSION THREAD

**Discussion by wgraaf** [AvaloniaUI/Avalonia#11907] [Q&A]
For (Repeat Last) Find-functionality in a datagrid, I need to iterate over the items in (potentially) sorted order.
The Items collection always stays in 'natural' or initial order.
The SelectedIndex property does reflect the sort order, but I don't know how to get the next item (in the sorted list) from there.

How can this be accomplished?

**Comment by maxkatz6** [comment]
I don't think it's possible with a data grid control.

**Reply by timunie** [reply]
* At least not without reflection *. If you are 👿 , you can try to access the underlaying DataGridCollectionView and get the sorted items there iirc. Or you use something like `DynamicData` and handle Sorting and filtering on your own. Still no "easy" solution here.

**Comment by wgraaf** [comment]
reflection got the job done:

          var dataConnection = DataGrid?.GetType().GetProperty("DataConnection", BindingFlags.Instance | BindingFlags.NonPublic)?.GetValue(DataGrid, null);
          var dataGridCollectionView = dataConnection?.GetType().GetProperty("DataSource")?.GetValue(dataConnection, null) as DataGridCollectionView;

dataGridCollectionView[idx] or dataGridCollectionView.GetItemAt(idx) gets item by sorted index

**Comment by VMelnalksnis** [comment]
I guess that is only an issue if you bind items to something like a `List<>`. I added a typed wrapper around `DataGridCollectionView`, which I use for binding to `ItemsSource`. I just checked, the elements in it are in the displayed (sorted) order. FYI, I omitted `INotifyCollectionChanged` implementation as I don't think that it's relevant for this, but maybe it is.
```csharp
using System.Collections;
using System.Collections.Generic;
using System.Linq;

using Avalonia.Collections;

public sealed class DataGridItemCollectionView<T> : IEnumerable<T>
{
    private readonly DataGridCollectionView _dataGridCollectionView;

    public DataGridItemCollectionView(IEnumerable<T> source)
        : this(new DataGridCollectionView(source))
    {
    }

    public DataGridItemCollectionView(DataGridCollectionView dataGridCollectionView)
    {
        _dataGridCollectionView = dataGridCollectionView;
    }

    public static implicit operator DataGridCollectionView(DataGridItemCollectionView<T> view) =>
        view._dataGridCollectionView;

    public IEnumerator<T> GetEnumerator() => _dataGridCollectionView.Cast<T>().GetEnumerator();

    IEnumerator IEnumerable.GetEnumerator() => _dataGridCollectionView.GetEnumerator();
}
```

**Reply by wgraaf** [reply]
Thanks for the reply.
For my understanding: this would require changing the Avalonia DataGrid source code?
Or is there another way to make DataGrid use the DataGridCollectionView wrapper?

**Reply by VMelnalksnis** [reply]
There is no need to change the source code, I'm using it right now with latest v11 release, and it worked with v0.10 as well. In my view model (DataContext) I have
```csharp
public DataGridItemCollectionView<TRow> Rows { get; protected set; }
```
and in the view
```xml
<DataGrid
    ItemsSource="{Binding Rows, Mode=TwoWay}"
    SelectedItem="{Binding Selected, Mode=TwoWay}" SelectionMode="Single"
```
Beyond snipets, you can check out how I use it here - [view](https://github.com/VMelnalksnis/Gnomeshade/blob/d221ef3c8560317bf1d45a97b1c1cabee536bd78/source/Gnomeshade.Desktop/Views/Accesses/OwnerView.axaml#L19C18-L19C18) and [view model](https://github.com/VMelnalksnis/Gnomeshade/blob/d221ef3c8560317bf1d45a97b1c1cabee536bd78/source/Gnomeshade.Avalonia.Core/OverviewViewModel.cs#L41C2-L41C2).

**Reply by wgraaf** [reply]
Thanks. That's a lot to study.
Does this use DynamicData Tim mentioned above?

**Reply by VMelnalksnis** [reply]
No, just the code here, I only left out extra things I added for code reuse/IPropertyChanged source-generation. I also set custom filter/sorting from the [view model](https://github.com/VMelnalksnis/Gnomeshade/blob/4bea3b3e6e3d41dfc2cc4fa28749345fc2329d0a/source/Gnomeshade.Avalonia.Core/Transactions/TransactionViewModel.cs#L67C6-L67C6). But maybe I'm missing something since the maintainers said otherwise.
